Migrate gym to gymnasium (#6309)
Unity-Technologies/ml-agents: com.unity.ml-agents/Documentation~/Python-Gym-API.md

# Unity ML-Agents Gym Wrapper

> A common way in which machine learning researchers interact with simulation environments is via a wrapper provided by the Farama Foundation called `gymnasium` (formerly known as OpenAI `gym`). For more information on the gymnasium interface, see the [Gymnasium Documentation](https://gymnasium.farama.org/index.html) and [Gymnasium Github repository](https://github.com/Farama-Foundation/Gymnasium).

Reviewer: ```suggestion
A common way in which machine learning researchers interact with simulation environments is via a wrapper provided by the Farama Foundation called `gymnasium` (formerly known as OpenAI `gym`). For more information on the gymnasium interface, visit the [Gymnasium Documentation](https://gymnasium.farama.org/index.html), and the [Gymnasium Github repository](https://github.com/Farama-Foundation/Gymnasium).
```

> We provide a gym wrapper and instructions for using it with existing machine learning algorithms which utilize gymnasium. Our wrapper provides interfaces on top of our `UnityEnvironment` class, which is the default way of interfacing with a Unity environment via Python.

Reviewer: ```suggestion
The ML Agents package provides a gym wrapper and instructions for using it with existing machine learning algorithms which utilize gymnasium. This wrapper provides interfaces on top of the `UnityEnvironment` class, which is the default way of interfacing with a Unity environment via Python.
```

- Environment registration for use with `gym.make()` is currently not supported.
- Calling env.render() will not render a new frame of the environment. It will return the latest visual observation if using visual observations.

## Training with Stable-Baselines3

> [Stable-Baselines3](https://github.com/DLR-RM/stable-baselines3) (SB3) is a set of reliable, actively maintained implementations of reinforcement learning algorithms in PyTorch. It is the community successor to OpenAI Baselines and, like this wrapper, is built on the Farama Foundation `gymnasium` API, so ML-Agents environments can be trained with it directly.

Reviewer: ```suggestion
[Stable-Baselines3](https://github.com/DLR-RM/stable-baselines3) (SB3) is a set of reliable, actively maintained implementations of reinforcement learning algorithms in PyTorch. It is the community successor to OpenAI Baselines and, like this wrapper, is built on the Farama Foundation `gymnasium` API, so you can train ML-Agents environments with it directly.
```

Install SB3 with:

```sh
pip install stable-baselines3
```

### Example - PPO

> To train an agent with PPO on a single-agent environment, create a file called `train_unity.py` with the following code. Then create an `/envs/` directory and build the environment to that directory. For more information on building Unity environments, see [here](Learning-Environment-Executable.md).

Reviewer: ```suggestion
To train an agent with PPO on a single-agent environment, create a file called `train_unity.py` with the following code. Then create an `/envs/` directory and build the environment to that directory. For more information on building Unity environments, refer to [Using an Environment Executable](Learning-Environment-Executable.md).
```
Matching link text to destination title

```python
from stable_baselines3 import PPO

from mlagents_envs.environment import UnityEnvironment
from mlagents_envs.envs.unity_gym_env import UnityToGymWrapper

def main():
    unity_env = UnityEnvironment(<path-to-environment>)
    env = UnityToGymWrapper(unity_env)
    model = PPO("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=100000)
    print("Saving model to unity_model.zip")
    model.save("unity_model")

if __name__ == "__main__":
    main()
```

To start the training process, run the following from the directory containing `train_unity.py`:

```sh
python train_unity.py
```

> Use the `"MlpPolicy"` for environments with vector observations and the `"CnnPolicy"` for environments with visual (image) observations.

Reviewer: ```suggestion
Use the `MlpPolicy` for environments with vector observations, and the `CnnPolicy` for environments with visual (image) observations.
```
We should avoid quotation marks in the docs where possible

### Example - DQN

DQN requires a discrete action space. For an environment with a single visual observation and branched discrete actions, enable `uint8_visual` (to output `uint8` image observations) and `flatten_branched` (to collapse the branched action space into a single `Discrete` space):

```python
from stable_baselines3 import DQN

from mlagents_envs.environment import UnityEnvironment
from mlagents_envs.envs.unity_gym_env import UnityToGymWrapper

def main():
    unity_env = UnityEnvironment(<path-to-environment>)
    env = UnityToGymWrapper(unity_env, uint8_visual=True, flatten_branched=True)
    model = DQN(
        "CnnPolicy",
        env,
        learning_rate=2.5e-4,
        buffer_size=50000,
        learning_starts=20000,
        target_update_interval=50,
        gamma=0.99,
        verbose=1,
    )
    model.learn(total_timesteps=1000000)
    print("Saving model to unity_model.zip")
    model.save("unity_model")

if __name__ == "__main__":
    main()
```

### Training on multiple environments in parallel

> SB3 can train on several environment instances at once using a vectorized environment. Each Unity instance must use a distinct `base_port` so the instances do not conflict:

Reviewer: ```suggestion
SB3 can train on several environment instances at once using a vectorized environment. Each Unity instance must use a distinct `base_port` so the instances don't conflict:
```

```python
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import SubprocVecEnv

from mlagents_envs.environment import UnityEnvironment
from mlagents_envs.envs.unity_gym_env import UnityToGymWrapper

def make_unity_env(env_directory, num_env, start_index=0):
    """Create a wrapped, vectorized Unity environment."""

    def make_env(rank):
        def _thunk():
            unity_env = UnityEnvironment(env_directory, base_port=5000 + rank)
            return UnityToGymWrapper(unity_env)

        return _thunk

    return SubprocVecEnv([make_env(i + start_index) for i in range(num_env)])

def main():
    env = make_unity_env(<path-to-environment>, 4)
    model = PPO("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=100000)

if __name__ == "__main__":
    main()
```

## Run Google Dopamine Algorithms

> > **Note:** The walkthrough below was written for an older, OpenAI `gym`-based release of Dopamine. This wrapper now follows the Farama Foundation `gymnasium` API: `reset()` returns `(observation, info)` and `step()` returns `(observation, reward, terminated, truncated, info)`. Recent versions of Dopamine target `gymnasium` and are compatible with the wrapper, but the exact file names, module paths, and configuration steps described here may differ from the version you install. Treat this section as a general guide rather than a step-by-step recipe.

Reviewer: ```suggestion
> [!NOTE] The following walkthrough was written for an older, OpenAI `gym`-based release of Dopamine. This wrapper now follows the Farama Foundation `gymnasium` API: `reset()` returns `(observation, info)` and `step()` returns `(observation, reward, terminated, truncated, info)`. Recent versions of Dopamine target `gymnasium` and are compatible with the wrapper, but the exact file names, module paths, and configuration steps described here may differ from the version you install. Treat this section as a general guide rather than a step-by-step recipe.
```
Adding formatting for callout box

First, open `dopamine/atari/run_experiment.py`. Alternatively, copy the entire `atari` folder, and name it something else (e.g. `unity`). If you choose the  copy approach, be sure to change the package names in the import statements in `train.py` to your new directory.

> Within `run_experiment.py`, we will need to make changes to which environment is instantiated, just as in the Stable-Baselines3 examples above. At the top of the file, insert

Reviewer: ```suggestion
Within `run_experiment.py`, we will need to make changes to which environment is instantiated, just as in the Stable-Baselines3 examples in the previous section. At the top of the file, insert
```
